apollo-federation/src/link/cost_spec_definition.rs

```diff
@@ -0,0 +1,165 @@
+use std::collections::HashMap;
+
+use apollo_compiler::ast::Argument;
+use apollo_compiler::ast::Directive;
+use apollo_compiler::name;
+use apollo_compiler::schema::Component;
+use apollo_compiler::schema::EnumType;
+use apollo_compiler::Name;
+use apollo_compiler::Node;
+use lazy_static::lazy_static;
+
+use crate::error::FederationError;
+use crate::link::spec::Identity;
+use crate::link::spec::Url;
+use crate::link::spec::Version;
+use crate::link::spec_definition::SpecDefinition;
+use crate::link::spec_definition::SpecDefinitions;
+use crate::schema::position::EnumTypeDefinitionPosition;
+use crate::schema::FederationSchema;
+
+pub(crate) const COST_DIRECTIVE_NAME_IN_SPEC: Name = name!("cost");
+pub(crate) const COST_DIRECTIVE_NAME_DEFAULT: Name = name!("federation__cost");
+
+pub(crate) const LIST_SIZE_DIRECTIVE_NAME_IN_SPEC: Name = name!("listSize");
+pub(crate) const LIST_SIZE_DIRECTIVE_NAME_DEFAULT: Name = name!("federation__listSize");
+
+#[derive(Clone)]
+pub(crate) struct CostSpecDefinition {
+    url: Url,
+    minimum_federation_version: Option<Version>,
+}
+
+impl CostSpecDefinition {
+    pub(crate) fn new(version: Version, minimum_federation_version: Option<Version>) -> Self {
+        Self {
+            url: Url {
+                identity: Identity::cost_identity(),
+                version,
+            },
+            minimum_federation_version,
+        }
+    }
+
+    pub(crate) fn cost_directive(
+        &self,
+        schema: &FederationSchema,
+        arguments: Vec<Node<Argument>>,
+    ) -> Result<Directive, FederationError> {
+        let name = self
+            .directive_name_in_schema(schema, &COST_DIRECTIVE_NAME_IN_SPEC)?
+            .unwrap_or(COST_DIRECTIVE_NAME_DEFAULT);
+
+        Ok(Directive { name, arguments })
+    }
+
+    pub(crate) fn list_size_directive(
+        &self,
+        schema: &FederationSchema,
+        arguments: Vec<Node<Argument>>,
+    ) -> Result<Directive, FederationError> {
+        let name = self
+            .directive_name_in_schema(schema, &LIST_SIZE_DIRECTIVE_NAME_IN_SPEC)?
+            .unwrap_or(LIST_SIZE_DIRECTIVE_NAME_DEFAULT);
+
+        Ok(Directive { name, arguments })
+    }
+
+    pub(crate) fn propagate_demand_control_directives(
+        &self,
+        subgraph_schema: &FederationSchema,
+        source: &apollo_compiler::ast::DirectiveList,
+        dest: &mut apollo_compiler::ast::DirectiveList,
+        original_directive_names: &HashMap<Name, Name>,
+    ) -> Result<(), FederationError> {
+        let cost_directive_name = original_directive_names.get(&COST_DIRECTIVE_NAME_IN_SPEC);
+        if let Some(cost_directive) = source.get(
+            cost_directive_name
+                .unwrap_or(&COST_DIRECTIVE_NAME_IN_SPEC)
+                .as_str(),
+        ) {
+            dest.push(Node::new(self.cost_directive(
+                subgraph_schema,
+                cost_directive.arguments.clone(),
+            )?));
```

Comment:
The first thing we're doing is getting the directive name from the supergraph schema, then we use that name to get the directive definition in the supergraph schema. If the supergraph schema has the cost directive defined, we propagate it to the subgraph, and the use of `self.cost_directive()` is going to ask the _subgraph_ what it wants to call the cost directive. More details re: the supergraph lookup on your other comment.

Reply:
thank you for the explanation! I’d still add this to doc comments/regular comments in case other people have questions about the feature. 

do we expect subgraphs to call cost directives something other than `@cost`? That doesn’t seem standard